axelor-contract/src/main/java/com/axelor/apps/contract/web/ContractLineController.java

```diff
@@ -59,4 +75,157 @@ public void createAnalyticDistributionWithTemplate(
             .createAnalyticDistributionWithTemplate(contractLine, contract);
     response.setValue("analyticMoveLineList", contractLine.getAnalyticMoveLineList());
   }
+
+  public void setAxisDomains(ActionRequest request, ActionResponse response) {
+    try {
+      Contract contract = null;
+      if (request.getContext().getParent() != null
+          && (Contract.class)
+              .equals(request.getContext().getParent().getParent().getContextClass())) {
+        contract = request.getContext().getParent().getParent().asType(Contract.class);
+      }
+
+      AnalyticToolService analyticToolService = Beans.get(AnalyticToolService.class);
+
+      for (int i = startAxisPosition; i <= endAxisPosition; i++) {
+        List<Long> analyticAccountList = new ArrayList<>();
+        if (contract != null
+            && analyticToolService.isPositionUnderAnalyticAxisSelect(contract.getCompany(), i)) {
+
+          AnalyticAxis analyticAxis = new AnalyticAxis();
+
+          for (AnalyticAxisByCompany axis :
+              Beans.get(AccountConfigService.class)
+                  .getAccountConfig(contract.getCompany())
+                  .getAnalyticAxisByCompanyList()) {
+            if (axis.getSequence() + 1 == i) {
+              analyticAxis = axis.getAnalyticAxis();
+            }
+          }
+
+          for (AnalyticAccount analyticAccount :
+              Beans.get(AnalyticAccountRepository.class).findByAnalyticAxis(analyticAxis).fetch()) {
+            analyticAccountList.add(analyticAccount.getId());
+          }
+
+          if (ObjectUtils.isEmpty(analyticAccountList)) {
+            response.setAttr(
+                "axis".concat(Integer.toString(i)).concat("AnalyticAccount"),
+                "domain",
+                "self.id IN (0)");
+          } else {
+            if (contract.getCompany() != null) {
+              String idList =
+                  analyticAccountList.stream()
+                      .map(Object::toString)
+                      .collect(Collectors.joining(","));
+
+              response.setAttr(
+                  "axis" + i + "AnalyticAccount",
+                  "domain",
+                  "self.id IN ("
+                      + idList
+                      + ") AND self.statusSelect = "
+                      + AnalyticAccountRepository.STATUS_ACTIVE
+                      + " AND (self.company is null OR self.company.id = "
+                      + contract.getCompany().getId()
+                      + ")");
+            }
+          }
+        }
+      }
```

Comment:
most of this method should be done inside a service, the controller should only be here to extract data from `request` and return the computed values of a service to the view using `response`